### ISSUE TITLE
[feature] Add a from_dict method for MultiIndex construction

### ISSUE THREAD

**Issue by mefergus** [pandas-dev/pandas#18470]
I've got a use case where I have a data frame with a ton of columns and I want to add a level to the column index to help in groupby operations.  My first instinct was to create a dictionary to illustrate the grouping, and use that pretty directly in indexing the dataframe. I.e.:

```python
hierarchy = {'g0':['c0','c1'], 'g1':['c2', 'c3', 'c4']}
new_index = MultiIndex.from_dict(hierarchy)
```

This just strikes me as a pretty natural way to think about hierarchical structure of data.

So something like this, but with support for nested dictionaries, added to the MultiIndex class:

```python
def from_dict(cls, data, names=None):
    '''
    Construct a two level MultiIndex from a dictionary-like object.

    Parameters
    ----------
    data : dict 
        {group : array-like}
        Here the array-like represents the group members
    names : optional names for the levels.

    Returns
    -------
    index : MultiIndex
	
    Notes
    -----
    Respects grouping order if dict is an OrderedDict
    '''
    lvl1 = []
    lvl2 = []
    for name, cols in data.items():
        lvl1 += [name] * len(cols)
        lvl2 += cols

    return MultiIndex.from_arrays([lvl1, lvl2], names=names)

```

**Comment by jreback** [comment]
```
In [50]: pd.MultiIndex.from_tuples([(name, l) for name, levels in hierarchy.items() for l in levels])
Out[50]: 
MultiIndex(levels=[['g0', 'g1'], ['c0', 'c1', 'c2', 'c3', 'c4']],
           labels=[[0, 0, 1, 1, 1], [0, 1, 2, 3, 4]])
```

The issue here is that a dictionary has an arbitrary ordering, so construction of the above is not deterministic. Sure you can do it, but I don't think this solves anything for the general case.

**Comment by mefergus** [comment]
I have been using `OrderedDict` to preserve ordering when constructing just the `MultiIndex`.  I've since decided it probably makes more sense to create a new `DataFrame` with columns selected from an existing one using the relationship I describe in the `OrderedDict`, and `concat`-ing them with the new labels as the keys. It's more flexible that way, too, since I don't need to grab all the columns, or I can put columns in multiple groups if that makes sense (which it does in some instances, for me).

I've got hundreds of columns that I want to group for summaries and needed a safe way to get the `labels` right (i.e. trying to avoid `labels` altogether) - I've got a relatively simple helper function doing that for me now, but I can see how it might not be a common need. I looked to working with the `MultiIndex` since that seemed to be where I needed the new information and I could modify in-place.

I still like to think about my hierarchy in a a dict-like fashion :)
